src/madness/world/worldrmi.cc

```diff
@@ -332,10 +335,19 @@ namespace madness {
     void RMI::RmiTask::huge_msg_handler(void *buf, size_t /*nbytein*/) {
         const size_t* info = (size_t *)(buf);
         int nword = HEADER_LEN/sizeof(size_t);
-        int src = info[nword];
-        size_t nbyte = info[nword+1];
-
-        RMI::task_ptr->hugeq.push_back(std::make_pair(src,nbyte));
+        const int src = info[nword];
+        const size_t nbyte = info[nword+1];
+        const int tag = info[nword+2];
+
+        // extra dose of paranoia: assert that we never process so many huge messages
+        // that the tag wraparound somewhere becomes possible ...
+        // the worst case is where only one node sends huge messages to every node in the communicator
+        // AND it has enough threads to use up all tags
+        // NB list::size() is O(1) in c++11, but O(N) in older libstdc++
+        MADNESS_ASSERT(ThreadPool::size() < RMI::RmiTask::unique_tag_period() ||
+                       RMI::task_ptr->hugeq.size() <
+                           RMI::RmiTask::unique_tag_period() / comm.Get_size());
```

Comment:
@evaleev I'm sorry but the check is incorrect. If the OR were changed to AND, then it would be better but still a bad check.

Let me insert some numbers to illustrate. Let
T = 64 = number of threads per process 
U = 4096 = number of unique tags
H = 0 = number of elements in huge message queue
P = 4097 = number of MPI processes

The Boolean equation is

T < U || H < U / P

Plug in the numbers and you get

64 < 4096 || 0 < 4096 / 4097

The only number that will change during the run is H. 
64 is always less than 4096, and thus the expression will always be true, no matter what H is.

Now if we correct the expression and put an AND in place of the OR, we get

64 < 4096 && 0 < 4096 / 4097

Here the left is still true, but the right is false (due to integer division, where 4096 / 4097 == 0). In this example, with no messages in the huge queue, the assert will fail.

Both cases are wrong. With the OR it can only fail if the number of threads is at an unrealistic value. With the AND, it will always fail for large process counts. The right hand cheek will also fail if

H= 2
P = 2048

64 < 4096 && 2 < 4096 / 2048

Here again reasonable values that should not be a failure.

If you want to keep the thread check, that is fine. I don't think it is necessary, but it's not going to cause bad behavior like the other issues.